### ISSUE TITLE
nspawn: cannot create block & char devices under /run/systemd/inaccessible

### ISSUE THREAD

**Issue by dongsupark** [systemd/systemd#4484]
### Submission type
- [x] Bug report
- [ ] Request for enhancement (RFE)
### systemd version the issue has been seen with

f70ebf1c ("Merge pull request #4474 from poettering/nsswitch"), 2016-10-25
: for both outside container and inside container
### Used distribution

Fedora 24
: for both outside container and inside container.
### In case of bug report: Expected behaviour you didn't see

(Inside a nspawn container) all 6 files should be created under `/run/systemd/inaccessible`.

```
-bash-4.3# ls -l /run/systemd/inaccessible/
total 0
b--------- 2 root root  0 Oct 24 17:37 blk
c--------- 2 root root  0 Oct 24 17:37 chr
d--------- 2 root root 40 Oct 24 17:37 dir
p--------- 1 root root  0 Oct 24 17:37 fifo
---------- 1 root root  0 Oct 24 17:37 reg
s--------- 1 root root  0 Oct 24 17:37 sock
```
### In case of bug report: Unexpected behaviour you saw

(Inside a nspawn container) only 4 files are created under `/run/systemd/inaccessible`, while both blk and chr are not created.

```
-bash-4.3# journalctl -b -o cat | grep inacc
Couldn't stat device /run/systemd/inaccessible/chr
Couldn't stat device /run/systemd/inaccessible/blk

-bash-4.3# ls -l /run/systemd/inaccessible/
total 0
d--------- 2 root root 40 Oct 24 17:37 dir
p--------- 1 root root  0 Oct 24 17:37 fifo
---------- 1 root root  0 Oct 24 17:37 reg
s--------- 1 root root  0 Oct 24 17:37 sock
> …
```
### In case of bug report: Steps to reproduce the problem

```
$ sudo systemd-nspawn -b -D /srv/container --capability=CAP_SYS_RAWIO,CAP_MKNOD
... (log in to the container, and run)
# ls -l /run/systemd/inaccessible/
```

AFAICT this happens because systemd-nspawn cannot create block & char devices in [mount_setup](https://github.com/systemd/systemd/blob/master/src/core/mount-setup.c#L415), returning `-EPERM`.
Inside the container, even running `"mknod /run/systemd/inaccessible/chr c 0 0"` does not work. ("permission denied")
Creating under /dev like `"mknod /dev/null2 c 1 3"` does not work either.
Even after retaining capabilities with an option like `"systemd-nspawn ... --capability=CAP_SYS_RAWIO,CAP_MKNOD"` it doesn't bring any change.

Of course it works fine outside the nspawn container.
Even inside container, for some users or for specific setup, it seems to work fine. But I haven't been able to address exactly what makes such difference.

See also https://github.com/systemd/systemd/pull/4473#issuecomment-255816636

/cc @evverx @tixxdz

**Comment by evverx** [comment]
> sudo systemd-nspawn -b -D /srv/container --capability=CAP_SYS_RAWIO,CAP_MKNOD

Did you pass the `-U` option?
`systemd-nspawn` should work fine without `-U`

**Comment by dongsupark** [comment]
@evverx I tried both with `-U` and without `-U`. It was the same.

**Comment by evverx** [comment]
What is the output of the

```
systemctl --version
```

and

```
cat /sys/fs/cgroup/devices/machine.slice/<your-machine>.scope/devices.list
```

on the host?

Are there

```
c 0:0 rwm
b 0:0 rwm
```

?

> systemd-nspawn should work fine without -U*
- Since [`v231`](https://github.com/systemd/systemd/commit/0d9e799102674c50e075) :-)

**Comment by dongsupark** [comment]
@evverx 
On the host,

```
$ systemctl --version
systemd 231
+PAM +AUDIT +SELINUX +IMA -APPARMOR +SMACK +SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT +GNUTLS +ACL +XZ -LZ4 +SECCOMP +BLKID +ELFUTILS +KMOD +IDN
```

```
$ cat /sys/fs/cgroup/devices/machine.slice/machine-container.scope/devices.list
c 1:3 rwm
c 1:5 rwm
c 1:7 rwm
c 1:8 rwm
c 1:9 rwm
c 5:0 rwm
c 5:2 rw
c 0:0 rwm
b 0:0 rwm
c 136:* rw
c 10:200 rwm
```

So both `"c 0:0 rwm"` and `"b 0:0 rwm"` do exist.
As noted above, I'm testing with systemd f70ebf1, which is newer than v231, for both host and container.

**Comment by evverx** [comment]
@dongsupark 

> As noted above, I'm testing with systemd f70ebf1, which is newer than v231

Oh, indeed, sorry.
Hm, I can't reproduce the `systemd (without -U)`-issue.
Just to be sure. What is the output of the

```
cat /proc/version
```

I have:

```
$ cat /proc/version
Linux version 4.7.9-200.fc24.x86_64 (mockbuild@bkernel01.phx2.fedoraproject.org) (gcc version 6.2.1 20160916 (Red Hat 6.2.1-2) (GCC) ) #1 SMP Thu Oct 20 14:26:16 UTC 2016
```

**Comment by dongsupark** [comment]
I revisited this issue. It consists of 2 different issues.
- Unable to create blk,chr under `/run/systemd/inaccessible`, for both user-namespace and non-user-namespace.

This issue has been already fixed. With release v232 and current master 375fd155, it works just fine without user namespace. OTOH with user namespace, it still cannot create blk,chr.
- Unable to create blk,chr with user namespace.

This issue still persists. That's actually because the Kernel does not allow blk and chr to be created inside user namespace. See [vfs_mknod() in Linux Kernel](https://kernel.googlesource.com/pub/scm/linux/kernel/git/stable/linux-stable/+/master/fs/namei.c#3672). There must be good reasons for doing that.

Thus we need to instead create inaccessible files from `outer_child()` in nspawn, when user namespace is enabled. I'll create a PR for doing that.

That's outcome from offline discussions with @tixxdz. Thanks!

**Comment by evverx** [comment]
We don't create `/run/systemd/inaccessible/{chr,blk}`
I think we should reopen this issue

**Comment by flokli** [comment]
@evverx I think same error here with systemd user sessions, `PrivateTmp` and `PrivateDevices` set to yes.

```systemd 234
+PAM -AUDIT -SELINUX -IMA -APPARMOR +SMACK -SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT +GNUTLS +ACL +XZ +LZ4 +SECCOMP +BLKID +ELFUTILS +KMOD -IDN2 +IDN default-hierarchy=hybrid
```

```
systemctl --user start foo
Aug 06 02:07:57 tp systemd[925]: Couldn't stat device /run/systemd/inaccessible/chr: Permission denied
Aug 06 02:07:57 tp systemd[925]: Couldn't stat device /run/systemd/inaccessible/blk: Permission denied
Aug 06 02:07:57 tp systemd[925]: Started foo
Aug 06 02:07:57 tp systemd[11924]: foo.service: Failed to drop capabilities: Operation not permitted 
Aug 06 02:07:57 tp systemd[925]: foo.service: Main process exited, code=exited, status=218/CAPABILITIES
Aug 06 02:07:57 tp systemd[925]: foo.service: Unit entered failed state.
Aug 06 02:07:57 tp systemd[925]: foo.service: Failed with result 'exit-code'.
```

User namespaces are disabled in the kernel (Archlinux default)

**Comment by evverx** [comment]
@flokli, `systemd` failed to drop capabilities. This seems to have to do with https://github.com/systemd/systemd/issues/4959 and https://github.com/systemd/systemd/issues/3944.

**Comment by poettering** [comment]
> We don't create /run/systemd/inaccessible/{chr,blk}
> I think we should reopen this issue

So I think it's entirely OK if we can't create these nodes. The code making use of these nodes already maps to almost-as-good /run/systemd/inaccessible/sock nodes in that case.

I mean, I think accepting fracefully that creating these device nodes doesn't always work is important. And if we accept that we have two options: create those nodes anyway, but as something else than char/block devices, so that all code using these nodes will transparently use the slightly incorrect but still good enough nodes. Or we handle this on the consumer side: if the nodes are missing the client code can just fall back to the slighly incorrect but still good enough nodes. The latter is implemented in b3d1d51603408e7aea7971fabf41b38c9e12fd69. 

hence I think we can close this now. I'll prep another PR that extends the comments about this situation, and I'll close this PR from that then.

**Comment by poettering** [comment]
The promised #8893 is posted now. Closing.
